### PR TITLE
[gitlab] Fix GitLab internal project visibility

### PR DESCRIPTION
**Pull request by pjlast**
Original Slack thread [here](https://sourcegraph.slack.com/archives/C05EMJM2SLR/p1687969384430849)

## Before this PR

We use a `min_access_level=20`  check on GitLab project queries, because `access_level` 10 is GUEST, and users with GUEST access cannot see repo contents, so we don't want to give access to those repos on Sourcegraph
HOWEVER, for Internal repos on GitLab, users can see the contents of the repo depending on how the repo is configured, regardless of their access level. So this puts us in a conundrum. If we have `min_access_level=20`, users can't see repos they should be able to see, but if we remove it, users can see repos they shouldn't be able to see

## After this PR

Adds a feature flag named "gitLabProjectVisibilityExperimental". When set to true:

@indradhanush  and I diffed some responses from the GitLab API, depending on whether or not you are part of the GitLab group or not, whether the project is visible or not, etc. etc. and there is a common field:
`default_branch` is only visible in the project API response if the user is able to see the contents of the project. If they cannot see the contents of the project, they cannot see the default branch of the project.
So by removing `min_access_level=20` and checking for the presence of `default_branch` on the returned project list, we can determine whether or not the user should be able to see the repo contents or not

## Test plan

VCR tests should be updated

### CODE REVIEW

**Review by indradhanush — APPROVED** [review]
LGTM as we paired on it. 😆

File: enterprise/internal/authz/gitlab/sudo.go [hunk 1]
```diff
@@ -243,7 +242,9 @@ func listProjects(ctx context.Context, client *gitlab.Client) (*authz.ExternalUs
 			}
 
 			for _, p := range projects {
-				projectIDs = append(projectIDs, extsvc.RepoID(strconv.Itoa(p.ID)))
+				if p.DefaultBranch != "" {
```
Comment (sashaostrikov) [inline enterprise/internal/authz/gitlab/sudo.go:245]: I'd add a nice comment that you had in Slack/PR here as well

**Review by sashaostrikov — APPROVED** [review]
LGTM! Let's add a comment directly in code?

File: enterprise/internal/authz/gitlab/sudo.go (hunk 1 above)
Comment (mrnugget) [reply enterprise/internal/authz/gitlab/sudo.go:245]: +500. This feels like a tripwire for future us. 

I'd also love to see a test for this change.

**Review by BolajiOlajide — APPROVED** [review]
+1 to @sashaostrikov's comment.

Asides that, this looks good to me.

File: enterprise/internal/authz/gitlab/sudo.go (hunk 1 above)
Comment (unknwon) [reply enterprise/internal/authz/gitlab/sudo.go:245]: +1024 really nice trick but even Cody won't be able to explain this 😂

File: internal/extsvc/gitlab/projects.go [hunk 2]
```diff
@@ -76,6 +76,14 @@ func (p Project) RequiresAuthentication() bool {
 	return p.Visibility == "private" || p.Visibility == "internal"
 }
 
+// ContentsVisible reports whether or not the repository contents of this project is visible to the user.
```
Comment (sashaostrikov) [inline internal/extsvc/gitlab/projects.go:79]: NICE!

**Review by kopancek — APPROVED** [review]

**Pull request merged**